### ISSUE TITLE
Create a specification for the command line interface

### ISSUE THREAD

**Issue by awvwgk** [fortran-lang/fpm#227]
To make clear what we expect from the command line interface we should create a specification of the current command line interface, similar to the specification for the manifest format. First, it allows to check for consistency for between the bootstrap and the Fortran version and gives a clear reference what to expect from an fpm implementation (of whatever kind). Also, it allows to propose new functionality independently of an implementation.

Good material might be the dub or cargo reference
- https://dub.pm/commandline
- https://doc.rust-lang.org/cargo/commands/index.html

The format should probably be man page compatible, preferably markdown, such that we can include it on the online documentation with various tools (FORD, jekyll, ...) and also translate it to roff. I personally prefer asciidoc to write man pages, which is translatable to roff and compatible with jekyll but probably not with FORD.

Any other opinion or suggestions on this?

**Comment by LKedward** [comment]
Yes this is definitely needed; we have two slightly divergent CLIs currently. I don't have experience with asciidoc, but I agree that it would be good to have man page compatibility for this. I don't see FORD compatibility as a necessity for this.

**Comment by urbanjost** [comment]
I use txt2man(1) markup myself so I have not tried asciidoc although I have heard about it. In my own environment the manpage document is written as text blocks in the Fortran source file and extracted and usuallly also converted to comments or a text variable declaration by a preprocessor which works the way I like (documentation, Fortran source all in one file and automatically converted to roff and then (via man2html or groff) converted to HTML) but I don't expect the rest of the world to do that.   

That being said, txt2man is widely available, simple, and works reasonably well for basic manpages. I keep meaning to make a Fortran program that is the equivalent but do not.

This is particularly apt as I was just going to open an issue and ask if we wanted to add a CLI option to specify custom compiler options like h-fpm recently added; but was thinking the syntax `fpm build -- -arg1 opt1 -arg2 -arg3... would be the easiest to implement and be a little easier to use but would be a different syntax for f-fpm than h-fpm. Wanted feedback on how people felt about it at all and what experiences everyone was having with the h-fpm version but the topics overlap; and at a minimum highlight the question of whether h-fpm and f-fpm need to stay in sync or not.

**Comment by urbanjost** [comment]
Note that `fpm help manual` writes out a file that was essentially written in `txt2man` markdown. Each help section was designed to be converted to roff via txt2man(1). The roff can then be converted to html, which can display on a github page if you turn on the doc section, or github markup allows you to put HTML into a `md` file by putting <? above the HTML; although a little tweaking was involved. I had put that file in the fpm repository for easier review of the documentation at one point.
It needs a little tweaking, but 

[mozilla.pdf](https://github.com/fortran-lang/fpm/files/5507892/mozilla.pdf)

**Comment by awvwgk** [comment]
@urbanjost Thanks for mentioning `txt2man`. This would be an alternative to consider. I mainly opened this issue to discuss and agree on tools for this purpose. I'm also happy to try something other than `asciidoc` for this purpose as well.

I wonder if `txt2man` is actually a cross-platform tool or if it is limited to Unix systems since it is written in bash/gawk (using GNU extensions)? I used `txt2man` a bit after you brought it up with the intrinsics module, the one thing I noticed is that it can become quite slow when converting several text files.

I want to decouple the Fortran or Haskell implementation from the specification on purpose. I think it makes more sense to document the wanted behaviour and than to provide the matching implementation. I would take the Fortran version as ground of truth for now.

> Wanted feedback on how people felt about it at all and what experiences everyone was having with the h-fpm version but the topics overlap; and at a minimum highlight the question of whether h-fpm and f-fpm need to stay in sync or not.

I would very much like them to stay in sync, having a common documentation and force to document any deviating behaviour might help with this.

> This is particularly apt as I was just going to open an issue and ask if we wanted to add a CLI option to specify custom compiler options like h-fpm recently added; but was thinking the syntax `fpm build -- -arg1 opt1 -arg2 -arg3... would be the easiest to implement

Please go ahead and open a new issue, I have a few suggestions for this.

**Comment by awvwgk** [comment]
I created #261 using markdown to write the man page and convert it using pandoc. There is also #230 using asciidoc as markup language and using asciidoctor to convert to roff and HTML. If there is interest I can create another PR with txt2man.

I would suggest that we review the PRs under consideration of:
- setup/usage of the toolchain
- the markup language to express the command line documentation
- the output (both HTML and roff/man page)
- the preview on GH

Any other point we should look out for?

**Comment by urbanjost** [comment]
I had something in mind that at some point we would have as part of a package a doc directory and that the install command could look at the suffix of the files and place at least common formats in an "installed" location.  Not really familiar with some of the other tools mentioned but I am assuming the processing would be done by the fpm developers to a few formats so users would not have to install a lot of software just to read the document.  At a minimum I want the help text of the program to be relatively complete and to contain links to WWW resources as appropriate, and for a manual to be availble in a PDF format. I am not quite sure if you are talking about documentation distributed as part of `fpm` for users or developer documentation or both? 

The help text was written specifically with it in mind that it could be converted to manpages and HTML and a PDF easilly, so I picked txt2man(1) because it is formatted completely with flat text and whitespace. So the direction is 'fpm help TOPIC|txt2man >TOPIC.1fpm or >fpm-TOPIC.1.  It seems like this would require the opposite process where the documents are maintained external to the program.  MSWindows (except in CygWin).  With fpm changing rapidly the potential for a mismatch between documentation and the version being used at this point seems high unless the documentation is in the code or distributed with each release. For now the simplest solution seemed to be to make sure at least basic documentation is built in.  Do other packages use a second application when help is called, or do they contain basic documentation built in?  The basic functionality I was modeling was largely based on cargo (and that is similar to what I normally do, so I am perhaps biased). Should all documentation or almost all go to a compantion program perhaps called fpm-help?  What do you see a new user having to do to access documentation? There not being any universal format for this I have pondered exactly what we can do so if you have a lot of dependencies in a program that you have documentation for the packages; ideally even on a platform not connected to a network or where users cannot install additional packages easily.

**Comment by awvwgk** [comment]
You give me quite a lot to think about. I will try to comment on the first part for now and come back later for the remainder, in reverse order.

> I am not quite sure if you are talking about documentation distributed as part of fpm for users or developer documentation or both?

I was thinking about user documentation which we can distribute with *fpm* once we have a fully functional `fpm-install` command.

> At a minimum I want the help text of the program to be relatively complete and to contain links to WWW resources as appropriate, and for a manual to be availble in a PDF format.

I agree. HTML and PDF is as far as I know something all three so far suggested tools can easily provide.

> Not really familiar with some of the other tools mentioned but I am assuming the processing would be done by the fpm developers to a few formats so users would not have to install a lot of software just to read the document.

Yes this is one of my main concern, whatever we choose should be immediately accessible by the user. The lowest barrier for interacting with the *fpm* documentation is currently visiting this repository via the GitHub web interface, therefore I think it is a great plus point if our approach allows a decent preview in the web interface by itself. Also important is how well the documentation works from the cloned git repository in a plain text editor, like nvim or emacs, or in an IDE, which might also provide a preview for certain formats. Also here if the format is reasonable readable without processing I consider this a great plus.

This becomes less of an issue once we have a dedicated place to display *fpm* specific documentation at https://fortran-lang.org or someplace else, because here we can easily show the processed output.

> I had something in mind that at some point we would have as part of a package a doc directory and that the install command could look at the suffix of the files and place at least common formats in an "installed" location.

This is definitely something I want to look into once we have included the `fpm-install` command. One possibility would be to extend the proposed `[install]` table with documentation or have a separate `[documentation]` table to specify the tools specify the sources and the tools required.
